### ISSUE TITLE
Menu hook doesn't work with retail TRAE exe

### ISSUE THREAD

**Issue by NicknineTheEagle** [TheIndra55/TRAE-menu-hook#9]
Menu hook doesn't appear to work with the original retail exe of Tomb Raider: Anniversary (which has SecuROM DRM), nothing happens when I try to boot the game. Only cracked exe from HATRED scene group works. The documentation makes no mention of needing to use a cracked exe so I assumed this isn't a known issue.

**Comment by TheIndra55** [comment]
Hi,

Thank you for opening this issue.

I test all releases with both the Steam and GOG retail executables. Would you be able to make sure you are using the latest version, and let me know which game version you have?

Thanks

**Comment by NicknineTheEagle** [comment]
As stated in the OP, I have disc version (European release). This game never received any updates to my knowledge.

**Comment by TheIndra55** [comment]
Sorry, do you have the latest version of Menu Hook?

I have code in place to only insert hooks after the game is done unpacking,
but I don't think I have tested these with SecureROM; only with Steam and
GOG.

**Comment by NicknineTheEagle** [comment]
I am using 2.0 menu hook release.

**Comment by NicknineTheEagle** [comment]
Tested https://github.com/TheIndra55/TRAE-menu-hook/releases/tag/v2.1, no changes.

**Comment by NicknineTheEagle** [comment]
Small note, if I hijacked a different dll with ASI Loader like winmm.dll, the game does start but Menu Hook doesn't actually work (intro not skipped, no watermark in main menu).

**Comment by TheIndra55** [comment]
Hi, sorry. I haven't looked into this yet since I currently do not hold a retail copy of the disc version of Anniversary.

**Comment by ThommyMD** [comment]
I have disc version (European release) too. The loader works for me but v1.1 only. Newer versions doesn't work, same for TR Legend Retail

**Comment by NicknineTheEagle** [comment]
> I have disc version (European release) too. The loader works for me but v1.1 only. Newer versions doesn't work, same for TR Legend Retail

There are patches for retail version? That's news to me, you got a link?

**Comment by TheIndra55** [comment]
Is v1.1 the last version that works, or do you have a more specific version?

**Comment by ThommyMD** [comment]
i've tested Menu hook 2.1 and 1.16., both doesn't work. With Menu hook 1.1 the mod files loads but there is no ingame menu (F8)

**Comment by NicknineTheEagle** [comment]
I've since had to deal with hooking SecuROM games myself and here are some comments on the issue.
https://github.com/TheIndra55/TRLAU-menu-hook/blob/9ca677499f129b79225cb0fc6ddebbc7573c3d65/src/Main.cpp#L9-L20
I do believe this hook gets triggered too early with SecuROM exe, before the game code is done unpacking. What you can do is test the memory here by looking for a specific pattern with Hooking.Patterns and further delaying the initialization if the check fails. Repeat every time the hook is triggered until the memory test passes.

**Comment by NicknineTheEagle** [comment]
https://github.com/TheIndra55/TRLAU-menu-hook/blob/9ca677499f129b79225cb0fc6ddebbc7573c3d65/src/Hook.cpp#L58-L59
This pattern is different in my exe, it ends with with 50 68. That said, I tried editing it but Hooking.Patterns can't find it for some reason even though I can literally see it in the memory using a debugger.

**Comment by TheIndra55** [comment]
Might SecuROM be relocating the code somewhere else? If i remember correctly Hooking.Patterns only scans the region of all image sections from the base address.

Don't get me wrong I'd love to support it, but I'm wondering if it's worth the effort when you can get a DRM-free EXE on GOG for $1. Especially since if they are relocating the code, my code still relies on hardcoded memory addresses in most places.

**Comment by NicknineTheEagle** [comment]
> Might SecuROM be relocating the code somewhere else? If i remember correctly Hooking.Patterns only scans the region of all image sections from the base address.

Ah, you seem to be right on that - the function that needs to be hooked is located at 0x014DF080 which is way beyond the original image sections. Any idea how to combat that?

**Comment by NicknineTheEagle** [comment]
I temporarily replaced pattern search with hardcoded address for the initial hook and that seems to have worked. No more relocated functions, I guess?

![Image](https://github.com/user-attachments/assets/a98bda3e-a467-4db3-91c1-ab01fecfcad6)

**Comment by TheIndra55** [comment]
Interesting, so was it being initialized too early or also the base address being different? Since the hook works seem like all functions and memory are still at the same address.

**Comment by NicknineTheEagle** [comment]
> Interesting, so was it being initialized too early or also the base address being different? Since the hook works seem like all functions and memory are still at the same address.

Neither, the issue was that A) the search pattern from `Hook::Initialize` slightly different in SecuROM exe (specified earlier) and B) the function hooked in `Hook::Initialize` is relocated by SecuROM so Hooking.Patterns can't find it by default. The former is an easy fix but I'm not quite sure how to get around the latter.

**Comment by TheIndra55** [comment]
Thanks you, the updated pattern is helpful and I'll see if I can repro this myself this week. I remember being able to specify the base address in Hooking.Patterns so perhaps I can hardcode it to `0x400000`.

**Comment by NicknineTheEagle** [comment]
> Thanks you, the updated pattern is helpful and I'll see if I can repro this myself this week. I remember being able to specify the base address in Hooking.Patterns so perhaps I can hardcode it to `0x400000`.

I don't believe base address is the problem, from what I'm seeing, Hooking.Patterns already takes base address from exe header. By default, it searches the memory in the range starting from the base address and ending with the end of exe file. Relocated functions are in the heap so you need to search that somehow.

**Comment by NicknineTheEagle** [comment]
Here's how Need for Speed: Undercover Generic Fix deals with SecuROM:
https://github.com/ThirteenAG/WidescreenFixesPack/blob/f04c6f682583600c2ef27b602b7a7846b30eec0c/source/NFSUndercover.GenericFix/dllmain.cpp#L966-L979
Perhaps, a similar solution can be implemented for Tomb Raider.

**Comment by NicknineTheEagle** [comment]
I tried doing this as a quick test and it worked:
```cpp
	uintptr_t RangeStart = 0, RangeEnd = 0;

	//securom compatibility
	uintptr_t ModuleStart = (uintptr_t)GetModuleHandle(NULL);
    MEMORY_BASIC_INFORMATION mbi;
    auto pattern = hook::pattern("FF 25"); //jmp
    for (size_t i = 0; i < pattern.size(); i++)
    {
		auto addr = *pattern.get(i).get<uintptr_t>(2);
        VirtualQuery((PVOID)addr, &mbi, sizeof(MEMORY_BASIC_INFORMATION));
        if (RangeEnd < addr && mbi.Protect == PAGE_EXECUTE_WRITECOPY && mbi.AllocationBase == (PVOID)ModuleStart)
        {
            RangeStart = (uintptr_t)mbi.BaseAddress;
            RangeEnd = (uintptr_t)mbi.BaseAddress + mbi.RegionSize;
        }
    }

#ifndef TR8
	//auto match = hook::pattern(RangeStart, RangeEnd, "A1 ? ? ? ? 8B 0D ? ? ? ? 68 ? ? ? ? 50 E8").count(1);
	auto match = hook::pattern(RangeStart, RangeEnd, "A1 ? ? ? ? 8B 0D ? ? ? ? 68 ? ? ? ? 50 68").count(1);
	s_deviceManager = *match.get_first<void*>(7);
#else
	auto match = hook::pattern("8B 0D ? ? ? ? 8B 01 8B 15 ? ? ? ? 8B 00 68").count(1);
	s_deviceManager = *match.get_first<void*>(2);
#endif
```
I'd submit a PR but I don't have the other two TR games set up.

**Comment by TheIndra55** [comment]
Note that the recent release doesn't fix this but I wanted to get another release out and wasn't able to setup a VM yet for the SecuROM stuff.
